File: 2022/01/31/longestCommonPrefix.swift
// https://leetcode.com/problems/longest-common-prefix/
// Author: Andrew Jarombek
// Date: 1/31/2022

func longestCommonPrefix(_ strs: [String]) -> String {
    var prefix = strs[0]

    for index in 1..<strs.count {
        let str = strs[index]

        for end in stride(from: min(str.count, prefix.count), to: -1, by: -1) {
            if end == 0 {
                return ""
            }

            let strIndex = str.index(str.startIndex, offsetBy: end)
            let prefixIndex = prefix.index(prefix.startIndex, offsetBy: end)

            if (str[str.startIndex..<strIndex] == prefix[prefix.startIndex..<prefixIndex]) {
                prefix = String(str[str.startIndex..<strIndex])
                break
            }
        }
    }

    return prefix
}

File: 2022/02/04/cpp/longestCommonPrefix.cpp
// https://leetcode.com/problems/longest-common-prefix/
// Author: Andrew Jarombek
// Date: 2/4/2022

#include <string>
#include <vector>

using namespace std;

string longestCommonPrefix(vector<string>& strs) {
    string prefix = strs[0];

    for (int i = 1; i < strs.size(); i++) {
        string str = strs[i];
        int l = 1;

        if (str.size() < prefix.size()) {
            prefix = prefix.substr(0, str.size());
        }

        while (l <= str.size()) {
            if (prefix.substr(0, l) != str.substr(0, l)) {
                if (l == 1)
                    return "";

                prefix = prefix.substr(0, l - 1);
                l = str.size() + 1;
            } else {
                l++;
            }
        }
    }

    return prefix;
}